### PR TITLE
Buildsystem: mkimage / get cleanup

### PR DESCRIPTION
**Pull request by antonlacon**
These are general cleanups for `mkimage` and the `get*` scripts. They cover:
Line indent
Backticks to $()
Quoting and braces/brackets/{}s for variables
Using build variables for path settings
Convert to die()

Testing by fetching/building an RPi2 `make image`. More testing probably needed as I don't think that it uses all the code paths here. For the Makefile change, the change is correct, but I'm not sure `make src-pkg` works, as I have no `.stamps` directory after building an image so it fails. Thus, I'm not sure how it is used currently.

### CODE REVIEW

File: scripts/mkimage [hunk 1]
```diff
@@ -10,327 +10,326 @@
...
-  show_error() {
-    echo "image: error happen..."
-    echo
-    cat "$SAVE_ERROR"
-    echo
-    cleanup
-    exit
-  }
+cleanup() {
+  echo "image: cleanup...\n"
```
Comment (MilhouseVH) [inline scripts/mkimage]: `echo -e` to interpret `\n`

File: scripts/mkimage [hunk 2]
```diff
@@ -10,327 +10,326 @@
...
+  rm -rf "${LE_TMP}"
+}
+
+show_error() {
+  echo "image: error happen...\n"
```
Comment (MilhouseVH) [inline scripts/mkimage]: `-e`

File: scripts/mkimage [hunk 3]
```diff
@@ -10,327 +10,326 @@
...
+UUID_1=$(date '+%d%m')
+UUID_2=$(date '+%M%S')
+FAT_SERIAL_NUMBER="${UUID_1}${UUID_2}"
+UUID_SYSTEM="${UUID_1}-${UUID_2}"
 
 # create an image
-  echo
-  echo "image: creating file $(basename $DISK)..."
-  dd if=/dev/zero of="$DISK" bs=1M count="$DISK_SIZE" conv=fsync >"$SAVE_ERROR" 2>&1 || show_error
+echo "\nimage: creating file $(basename $DISK)..."
```
Comment (MilhouseVH) [inline scripts/mkimage]: `-e`, `${DISK} `

File: scripts/get_archive [hunk 4]
```diff
@@ -1,39 +1,41 @@
...
-while [ $NBWGET -gt 0 ]; do
-  for url in "$PKG_URL" "$PACKAGE_MIRROR"; do
-    rm -f $PACKAGE
-    if $WGET_CMD "$url"; then
-      CALC_SHA256="$(sha256sum $PACKAGE | cut -d" " -f1)"
+while [ ${NBWGET} -gt 0 ]; do
+  for url in "${PKG_URL}" "${PACKAGE_MIRROR}"; do
+    rm -f "${PACKAGE}"
+    if ${WGET_CMD} "${url}"; then
+      CALC_SHA256="$(sha256sum ${PACKAGE} | cut -d" " -f1)"
```
[the remark this answers is not in the document]
Comment (antonlacon) [reply scripts/get_archive]: I thought the quotes around the subshell would take care of it. Turns out they don't; they're just extra.

File: scripts/get_git [hunk 5]
```diff
@@ -36,104 +38,103 @@ _get_repo_clean() {
...
-  if [ -d "$d/.git" ]; then
+for d in "${SOURCES}/${1}/${1}-"* ; do
+  if [ -d "${d}/.git" ]; then
     if [ "${GIT_FOUND}" = "no" ]; then
-      cd $d
-      if [ "$PKG_URL" = "$(git remote get-url origin)" ]; then
-        if [ -n "$PKG_GIT_CLONE_BRANCH" -a $(git branch | grep "^\* ${PKG_GIT_CLONE_BRANCH}$" | wc -l) -eq 1 -o -z "$PKG_GIT_CLONE_BRANCH" ]; then
+      cd "${d}"
+      if [ "${PKG_URL}" = "$(git remote get-url origin)" ]; then
+        if [ -n "${PKG_GIT_CLONE_BRANCH}" -a $(git branch | grep "^\* ${PKG_GIT_CLONE_BRANCH}$" | wc -l) -eq 1 -o -z "${PKG_GIT_CLONE_BRANCH}" ]; then
```
[the remark this answers is not in the document]
Comment (antonlacon) [reply scripts/get_git]: I believe you're right, which makes the other elif tests further down checking if PKG_GIT_CLONE_BRANCH isn't null redundant.

**Comment by antonlacon** [comment]
Comments thus far should all be addressed.

File: scripts/get_archive [hunk 6]
```diff
@@ -2,38 +2,38 @@
...
-lock_source_dir $1
+lock_source_dir "${1}"
 
 # Check again in case of concurrent access - if nothing needs to be downloaded, exit now...
 _get_file_already_downloaded && exit 0
 
 # At this point, we need to download something...
-printf "%${BUILD_INDENT}c $(print_color CLR_GET "GET")      $1 (archive)\n" ' '>&$SILENT_OUT
+printf "%${BUILD_INDENT}c $(print_color CLR_GET "GET")      ${1} (archive)\n" ' '>&$SILENT_OUT
 export BUILD_INDENT=$((${BUILD_INDENT:-1}+$BUILD_INDENT_SIZE))
```
Comment (MilhouseVH) [inline scripts/get_archive]:
```
-export BUILD_INDENT=$((${BUILD_INDENT:-1}+$BUILD_INDENT_SIZE))
+export BUILD_INDENT=$((${BUILD_INDENT:-1}+${BUILD_INDENT_SIZE}))
```

File: scripts/mkimage [hunk 7]
```diff
@@ -10,327 +10,326 @@
...
-    [ -n "$SUDO_USER" ] && chown $SUDO_USER: ${DISK_BASENAME}.ova
-  fi
+if [ "${PROJECT}" = "Generic" ]; then
+  echo "image: creating open virtual appliance..."
+  # duplicate $DISK so anything we do to it directly doesn't effect original
+  dd if="${DISK}" of="${DISK_BASENAME}.tmp" bs=1M >"${SAVE_ERROR}" 2>&1 || show_error
+  # change syslinux default to 'run'
+  echo "image: modifying fs on part1 for open virtual appliance..."
+  sed -i "/DEFAULT/ s/installer/run/" "${LE_TMP}/syslinux.cfg"
+  sed -i "/set default=/s/\"Installer\"/\"Run\"/" "${LE_TMP}/grub.cfg"
```
Comment (MilhouseVH) [inline scripts/mkimage]: Can we change these two to be of the form `sed -e "<script>" -i filename`?

**Comment by MilhouseVH** [comment]
Many thanks for the udpates - just a couple more minors.

Suggestion for this PR: I think we can simplify [`lock_source_dir()`](https://github.com/LibreELEC/LibreELEC.tv/blob/99cc6922b4522e0cf03d222e2c11a1503cd60171/scripts/get#L16-L26) as follows:
```
lock_source_dir() {
  exec 99<"${SOURCES}/${1}"
  if ! flock --nonblock --exclusive 99; then
    echo "Project/Device ${DEVICE:-${PROJECT}} waiting, to avoid concurrent processing of ${1}..."
    flock --exclusive 99
  fi
}
```
What do you think? I ran some quick simulated tests and the simplified version worked as expected.

Suggestion for a separate PR: I'd like to see if we can move the `printf "%${BUILD_INDENT}c $(print_color CLR_GET "GET")` statements to a helper function, perhaps something like:
```
# p1: CLR_var; p2: p1 label; p3: text
build_progress() {
  printf "%${BUILD_INDENT}c $(print_color ${1} "${2}")      ${3}\n" ' '>&$SILENT_OUT
}
build_progress_inc() {
  build_progress "${1}" "${2}" "${3}"
  export BUILD_INDENT=$((${BUILD_INDENT:-1}+${BUILD_INDENT_SIZE}))
}
```
and then:
```
-printf "%${BUILD_INDENT}c $(print_color CLR_GET "GET")      ${1} (archive)\n" ' '>&$SILENT_OUT
-export BUILD_INDENT=$((${BUILD_INDENT:-1}+$BUILD_INDENT_SIZE))
+build_progress_inc CLR_GET "GET" "${1} (archive)"
```
etc. What do you think? Let me know if you want to tackle this or I can investigate it a bit further and push a PR if it pans out! Thanks.

**Comment by antonlacon** [comment]
The minors are in.

The `lock_source_dir()` change looks right to me. My understanding is: `exec` is opening up a file descriptor, `99`, pointing to the directory the source tarball is being downloaded to. It then checks if the descriptor is locked, and goes ahead to lock it if it isn't. The lock is released when the shell is closed (which is when the file is downloaded it moves onto the next `get` invocation or whatever other step is next). This should match the old behaviour.

I can look at the printf changes later this week, if you don't beat me to it. Probably Friday-ish.

**Comment by MilhouseVH** [comment]
> The minors are in.

Thanks! I'll add this PR to the nightly builds so that it can be tested for a few days, I'll also try some clean builds and report any issues.

> My understanding is: 

Yep, that's my understanding also.

> I can look at the printf changes later this week, if you don't beat me to it. Probably Friday-ish.

Great thanks. I haven't looked into it in any detail, but I think the kind of change I proposed _should_ work. I suggested using two functions (one that increments the indent, and one that doesn't) rather than use a single function with an extra "increment or don't increment" argument, as the two function solution just seemed slightly more sane. I suppose a single function with optional 4th argument can work, I just don't know what the argument value should be... `"increment"` (risk of typo)? `"yes"`? At least encoding the requirement in the function name ensures an error if there is a typo etc.

**Comment by MilhouseVH** [comment]
Thanks!

**Pull request merged**